File: backend/models/weather.py
from __future__ import annotations

from datetime import datetime

from sqlalchemy import DateTime, Float, ForeignKey, Integer, String, func
from sqlalchemy.dialects.sqlite import JSON
from sqlalchemy.orm import Mapped, mapped_column, relationship

from backend.database import Base


class WeatherSnapshot(Base):
    __tablename__ = "weather_snapshots"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    activity_id: Mapped[int | None] = mapped_column(
        Integer, ForeignKey("activities.id", ondelete="SET NULL"), unique=True
    )
    temp_c: Mapped[float | None] = mapped_column(Float)
    feels_like_c: Mapped[float | None] = mapped_column(Float)
    humidity: Mapped[float | None] = mapped_column(Float)
    wind_speed: Mapped[float | None] = mapped_column(Float)
    wind_gust: Mapped[float | None] = mapped_column(Float)
    wind_deg: Mapped[int | None] = mapped_column(Integer)
    conditions: Mapped[str | None] = mapped_column(String)
    description: Mapped[str | None] = mapped_column(String)
    pressure: Mapped[float | None] = mapped_column(Float)
    uv_index: Mapped[float | None] = mapped_column(Float)
    raw_data: Mapped[dict | None] = mapped_column(JSON)
    created_at: Mapped[datetime] = mapped_column(DateTime, server_default=func.now())

    activity = relationship("Activity", back_populates="weather")

File: frontend/src/api/weather.ts
import { fetchJson, fetchOptionalJson } from "./http";

export interface WeatherSnapshot {
  id: number;
  activity_id: number;
  temp_c: number | null;
  feels_like_c: number | null;
  humidity: number | null;
  wind_speed: number | null;
  wind_gust: number | null;
  wind_deg: number | null;
  conditions: string | null;
  description: string | null;
  pressure: number | null;
  uv_index: number | null;
  created_at: string | null;
  /**
   * Full OpenWeatherMap payload. Only present when ``?raw=true`` was
   * passed; needed to render the weather icon (``data[0].weather[0].icon``).
   */
  raw_data?: Record<string, unknown> | null;
}

export interface WeatherBackfillResult {
  enriched: number;
  skipped: number;
  failed: number;
  remaining: number;
  batch: number;
  dry_run: boolean;
  configured: boolean;
}

/**
 * Fetch the weather snapshot joined to ``activityId``.
 *
 * Returns ``null`` if the activity has no snapshot yet (backend 404).
 * Other errors propagate as thrown ``Error``.
 */
export async function getActivityWeather(
  activityId: number,
  options: { raw?: boolean } = {}
): Promise<WeatherSnapshot | null> {
  const qs = options.raw ? "?raw=true" : "";
  return fetchOptionalJson<WeatherSnapshot>(
    `/activities/${activityId}/weather${qs}`
  );
}

/**
 * Kick off a weather-enrichment pass on the backend.
 */
export async function backfillWeather(
  params: { batch?: number; dry_run?: boolean } = {}
): Promise<WeatherBackfillResult> {
  return fetchJson<WeatherBackfillResult>("/weather/backfill", {
    method: "POST",
    body: JSON.stringify({
      batch: params.batch ?? 50,
      dry_run: params.dry_run ?? false,
    }),
  });
}
